utility/DateTime.h

```diff
@@ -0,0 +1,157 @@
+#ifndef _DATETIME_H_
+#define _DATETIME_H_
+
+#include "TimeSpan.h"
+#include <Arduino.h>
+
+/** Constants */
+#define SECONDS_PER_DAY 86400L ///< 60 * 60 * 24
+#define SECONDS_FROM_1970_TO_2000                                              \
+  946684800 ///< Unixtime for 2000-01-01 00:00:00, useful for initialization
+
+uint8_t getDaysInMonth(uint16_t year, uint8_t month);
+
+/**************************************************************************/
+/*!
+    @brief  Simple general-purpose date/time class (no TZ / DST / leap
+            seconds).
+
+    This class stores date and time information in a broken-down form, as a
+    tuple (year, month, day, hour, minute, second). The day of the week is
+    not stored, but computed on request. The class has no notion of time
+    zones, daylight saving time, or
+    [leap seconds](http://en.wikipedia.org/wiki/Leap_second): time is stored
+    in whatever time zone the user chooses to use.
+
+    The class supports dates in the range from 1 Jan 2000 to 31 Dec 2099
+    inclusive.
+*/
+/**************************************************************************/
+class DateTime {
```

Comment:
Add support for RTCs which may not support certain functions like dates? Some way to query or report back unsupported.

Reply:
Don't RTCs track the day, month, and year by definition? If not, then this is best left for another pull request. My recommendation is to make a `TimeOfDay` class that only stores the time and day and write a `timeOfDay()` method for the RTC classes. This can be implemented in the RTC base class if needed.

Reply:
I want to say I have seen ones that do not support it. I just looked around a little it looks like those are grouped with RTCs but are more likely counters used for time tracking and watchdog functions. However one could use these to make a RTC. 

I think the DateTime interface could be made flexible to avoid this. For the most part many RTCs do support it. It is a gray area. Some have day of the week while others do not also. However that could be viewed as out of scope for DateTime, but a nice feature perhaps.

The DS1672 is grouped with RTC and classified as RTC. However is called binary counter RTC. It also looks like these parts are being phased out. For the most part many do have them, and it could be fair to say that if having to build some of the DateTime they build all of it.

Here is a good list of RTCs. I did not look at it too much recently. I remember looking back at it years ago.
https://www.digikey.com/en/products/filter/clock-timing-real-time-clocks/690

Reply:
`DateTime` is a time representation in broken-down form (year, month, day, hour, minute, second), similar to libc's `struct tm`. The RTC you cite as an example (DS1672) provides instead scalar time (similar to `time_t`). I don't see the point of adding scalar time to `DateTime`. The class already provides the means to convert to and from scalar time.

Reply:
Another one that is a little different is DS1683, which does not work like time_t exactly. It technically can count up to 34 years worth of time, which means despite being a 32 bit counter it only has one quarter the resolution of scalar time. Now all scalar time could be converted. This is true for roughly any counter size, as in this case the hardware is just an accurate time source.

It is correct that most if not all RTCs technically do support clock and calendar functions. Some support day of the week, leap year, daylight savings, alarm functions, EEPROM, etc. Not all RTCs support the same features. Some of these features are not part of RTC's or DateTime's scope. One thought is that the driver class could implement multiple base class interfaces.

Now DateTime could also make the policy that certain features are not supported depending on implementation of driver which returns a flag accordingly. Simple but annoying.

Or on the flip side it purposefully ignores certain things and forces the application code to correct DateTime. All while requiring the DateTime to pick up in software any missing hardware component found in RTC. Therefore all of DateTime is required without exception.

DateTime will likely need to use a notion of leap year and day light at some point. In this case you are really just converting scalar time into a format. So these functions would still be used by the RTC, however the interface for them is not present to my knowledge in RTC or DateTime. DateTime is the universal abstraction of time representation so there could be an argument for it there.

Unless the plan is for the user to be completely aware that they are using one RTC versus another. In which case they cannot benefit from objects. In this case they should be kept independent and this library should not exist.

I think I have made my point, so I am going to stop here.

Reply:
Maybe I'm dumb, but I honestly do not understand the point you are trying to make. The DS1683 is irrelevant to `DateTime`, as it is just another scalar time format. Do you know of any RTC that provide time in broken-down form as (hour, minute, second) but without (year, month, day)?

If you believe the `DateTime` class could benefit from a revamp, then you should open an issue, lay down your proposed additions to its interface (all new methods, their parameters, return values...), and state whether you would be willing to implement them. Otherwise we are just discussing in the vague.